AP_Mount: Solo gimbal params use HAVE_PAYLOAD_SPACE

Pedals2Paddles opened the pull request:
Adds the HAVE_PAYLOAD_SPACE check to the solo gimbal's driver.

WickedShell commented on libraries/AP_Mount/SoloGimbal_Parameters.cpp:
```diff
@@ -115,7 +115,9 @@ void SoloGimbal_Parameters::set_param(gmb_param_t param, float value) {
     strncpy(tmp_name, get_param_name(param), MAVLINK_MSG_PARAM_SET_FIELD_PARAM_ID_LEN);
 
-    mavlink_msg_param_set_send(_chan, 0, MAV_COMP_ID_GIMBAL, tmp_name, _params[param].value, MAV_PARAM_TYPE_REAL32);
+    if (HAVE_PAYLOAD_SPACE(_chan, GIMBAL_CONTROL)) {
```
You need to be checking `PARAM_SET` not `GIMBAL_CONTROL`.

WickedShell commented on libraries/AP_Mount/SoloGimbal_Parameters.cpp:
```diff
@@ -126,7 +128,7 @@ void SoloGimbal_Parameters::update()
     // retry initial param retrieval
     if(!received_all()){
-        if (tnow_ms-_last_request_ms > _retry_period) {
+        if ((tnow_ms-_last_request_ms > _retry_period) && (HAVE_PAYLOAD_SPACE(_chan, GIMBAL_CONTROL))) {
```
You need to be checking `PARAM_SET` not `GIMBAL_CONTROL`.

Pedals2Paddles commented:
A little too fast on the copy/paste. Fixed.

magicrub commented on libraries/AP_Mount/SoloGimbal_Parameters.cpp:
```diff
@@ -115,7 +115,9 @@ void SoloGimbal_Parameters::set_param(gmb_param_t param, float value) {
     strncpy(tmp_name, get_param_name(param), MAVLINK_MSG_PARAM_SET_FIELD_PARAM_ID_LEN);
 
-    mavlink_msg_param_set_send(_chan, 0, MAV_COMP_ID_GIMBAL, tmp_name, _params[param].value, MAV_PARAM_TYPE_REAL32);
+    if (HAVE_PAYLOAD_SPACE(_chan, PARAM_SET)) {
```
are you supposed to be setting _last_request_ms if there's no space and it doesn't send?

magicrub commented on libraries/AP_Mount/SoloGimbal_Parameters.cpp:
```diff
@@ -126,7 +128,7 @@ void SoloGimbal_Parameters::update()
     // retry initial param retrieval
     if(!received_all()){
-        if (tnow_ms-_last_request_ms > _retry_period) {
+        if ((tnow_ms-_last_request_ms > _retry_period) && (HAVE_PAYLOAD_SPACE(_chan, PARAM_REQUEST_LIST_SEND))) {
```
compile is failing. To fix, change:
`PARAM_REQUEST_LIST_SEND`
to
`PARAM_REQUEST_LIST`

Pedals2Paddles commented:
Also fixed what I believe was a race condition where parameter request and set were both using `_last_request_ms`. Added a `_last_set_ms` variable for the set.

peterbarker commented on libraries/AP_Mount/SoloGimbal_Parameters.cpp:
```diff
@@ -115,9 +116,11 @@ void SoloGimbal_Parameters::set_param(gmb_param_t param, float value) {
     strncpy(tmp_name, get_param_name(param), MAVLINK_MSG_PARAM_SET_FIELD_PARAM_ID_LEN);
 
-    mavlink_msg_param_set_send(_chan, 0, MAV_COMP_ID_GIMBAL, tmp_name, _params[param].value, MAV_PARAM_TYPE_REAL32);
+    if (HAVE_PAYLOAD_SPACE(_chan, PARAM_SET)) {
```
Usually we would return-early rather than enclose the call in the if block.  In this case you'd have to move the `_last_set_ms` assignment up.

peterbarker commented on libraries/AP_Mount/SoloGimbal_Parameters.cpp:
```diff
@@ -128,22 +131,28 @@ void SoloGimbal_Parameters::update()
-                _params[i].fetch_attempts++;
+        if ((_params[i].state == GMB_PARAMSTATE_ATTEMPTING_TO_SET) && (tnow_ms - _last_set_ms > _retry_period)) {
+            _last_set_ms = AP_HAL::millis();
+            if (HAVE_PAYLOAD_SPACE(_chan, PARAM_SET)) {
```
This is never going to become true once it becomes false.  Just break out of the loop if you run out of space.

peterbarker commented:
@magicrub Have you concerns been addressed here?

magicrub commented:
looks fine now, thanks!

rmackay9 commented:
I'll leave it to @peterbarker to merge this one I think.

Pedals2Paddles commented:
Woo, my list keeps getting shorter this week!
